cfspeedtest/cloudflare.py

```diff
@@ -0,0 +1,235 @@
+            coll.server.append(
+                float(r.headers["Server-Timing"].split("=")[1]) / 1e3
+            )
+            coll.request.append(
+                r.elapsed.seconds + r.elapsed.microseconds / 1e6
+            )
+        return coll
+
+    def sprint(self, label: str, result: TestResult) -> None:
+        """Add an entry to the suite results and log it."""
+        if self.logger:
+            self.logger.info("%s: %s", label, result.value)
+        self.results[label] = result
+
+    @staticmethod
+    def calculate_percentile(data: list[float], percentile: float) -> float:
+        """Find the percentile of a list of values."""
+        data = sorted(data)
+        idx = (len(data) - 1) * percentile
+        rem = idx % 1
+
+        if rem == 0:
+            return data[int(idx)]
+
+        edges = (data[int(idx)], data[int(idx) + 1])
+        return edges[0] + (edges[1] - edges[0]) * rem
+
+    def run_all(self) -> dict[str, TestResult]:
+        """Run the full test suite."""
+        meta = self.metadata()
+        self.sprint("ip", TestResult(meta.ip))
+        self.sprint("isp", TestResult(meta.isp))
+        self.sprint("location_code", TestResult(meta.location_code))
+        self.sprint("location_city", TestResult(meta.city))
+        self.sprint("location_region", TestResult(meta.region))
+
+        data = {"down": [], "up": []}
+        for test in self.tests:
+            timers = self.run_test(test)
+
+            if test.name == "latency":
+                latencies = [
+                    (timers.request[i] - timers.server[i]) * 1e3
+                    for i in range(len(timers.request))
+                ]
+                jitter = statistics.median(
+                    [
+                        abs(latencies[i] - latencies[i - 1])
+                        for i in range(1, len(latencies))
+                    ]
+                )
+                self.sprint(
+                    "latency",
+                    TestResult(round(statistics.median(latencies), 2)),
+                )
+                self.sprint("jitter", TestResult(round(jitter, 2)))
+                continue
+
+            if test.type == TestType.Down:
+                speeds = [
+                    int(test.bits / (timers.full[i] - timers.server[i]))
+                    for i in range(len(timers.full))
+                ]
+            else:
+                speeds = [
+                    int(test.bits / server_time)
+                    for server_time in timers.server
+                ]
+            data[test.type.name.lower()].extend(speeds)
+            self.sprint(
+                f"{test.name}_{test.type.name.lower()}_bps",
+                TestResult(int(statistics.mean(speeds))),
+            )
+        for k, v in data.items():
+            self.sprint(
+                f"90th_percentile_{k}_bps",
+                TestResult(int(self.calculate_percentile(v, 0.9))),
+            )
```

Comment:
This part produces an error when I remove a test type from tests provided to CloudflareSpeedtest.

For example, I removed the UPLOAD_TESTS, and that resulted in an IndexError.

Reply:
on the subject of this, it is strange to have keys that may or may not be present at the top level. it feels inconvenient to make people scan the whole dictionary, maybe we should move these under an inner dict? like `{"results": {...}}`

Reply:
Good point, sounds sensible to me.